TST: TestFunctionHelpersCompleteness.test_all_included fails with numpy-dev

New failure today in the devdeps job. But why would the order matter? Did I read the log wrong? cc @mhvk 🙏 

https://github.com/astropy/astropy/actions/runs/6484236887/job/17607622657?pr=15462

```
______________ TestFunctionHelpersCompleteness.test_all_included _______________

self = <astropy.units.tests.test_quantity_non_ufuncs.TestFunctionHelpersCompleteness object at ...>

    @needs_array_function
    def test_all_included(self):
        included_in_helpers = (
            SUBCLASS_SAFE_FUNCTIONS
            | UNSUPPORTED_FUNCTIONS
            | set(FUNCTION_HELPERS.keys())
            | set(DISPATCHED_FUNCTIONS.keys())
        )
>       assert all_wrapped == included_in_helpers
E       assert {<function ca...ac26030>, ...} == {<function tr...11ee030>, ...}
...
E         Full diff:
E           {
...
E            <function multi_dot at 0x7f46211cb2f0>,
E         +  <function diagonal at 0x7f46211cb4f0>,
E            <function meshgrid at 0x7f46211cb6b0>,
E         +  <function trace at 0x7f46211cb730>,
E            <function histogram_bin_edges at 0x7f46211d7b70>,
...
E            <function squeeze at 0x7f4621a07c30>,
E         -  <function diagonal at 0x7f4621a07df0>,
E         -  <function trace at 0x7f4621a14030>,
E            <function ravel at 0x7f4621a14230>,
...
E           }
astropy/units/tests/test_quantity_non_ufuncs.py:2471: AssertionError
```

--- a/astropy/units/quantity_helper/function_helpers.py
+++ b/astropy/units/quantity_helper/function_helpers.py
@@ -102,6 +102,9 @@ SUBCLASS_SAFE_FUNCTIONS |= {np.median}
 if NUMPY_LT_2_0:
     # functions removed in numpy 2.0; alias for np.round in NUMPY_LT_1_25
     SUBCLASS_SAFE_FUNCTIONS |= {np.msort, np.round_}  # noqa: NPY003
+else:
+    # Array-API compatible versions (matrix axes always at end).
+    SUBCLASS_SAFE_FUNCTIONS |= {np.linalg.diagonal, np.linalg.trace}
 
 # Implemented as methods on Quantity:
 # np.ediff1d is from setops, but we support it anyway; the others

--- a/astropy/units/tests/test_quantity_non_ufuncs.py
+++ b/astropy/units/tests/test_quantity_non_ufuncs.py
@@ -44,12 +44,19 @@ def get_wrapped_functions(*modules):
     else:
         from numpy.testing.overrides import allows_array_function_override
 
-    return {
-        name: f
-        for mod in modules
-        for name, f in mod.__dict__.items()
-        if callable(f) and allows_array_function_override(f)
-    }
+    wrapped_functions = {}
+    for mod in modules:
+        for name, f in mod.__dict__.items():
+            if callable(f) and allows_array_function_override(f):
+                # Indexing by just the name is easiest for test writing,
+                # but in numpy 2.0, there are 2 versions of diagonal and
+                # trace, one in the main namespace, and one in np.linalg.
+                # So, we distinguish those by adding the module name.
+                if name in wrapped_functions:
+                    name = mod.__name__.replace("numpy.", "") + "_" + name
+                wrapped_functions[name] = f
+
+    return wrapped_functions
 
 
 all_wrapped_functions = get_wrapped_functions(
@@ -2052,7 +2059,7 @@ class TestFFT(InvariantUnitTestSetup):
         self.check(np.fft.ifftshift)
 
 
-class TestLinAlg(metaclass=CoverageMeta):
+class TestLinAlg(InvariantUnitTestSetup, metaclass=CoverageMeta):
     def setup_method(self):
         self.q = (
             np.array([[1.0, -1.0, 2.0], [0.0, 3.0, -1.0], [-1.0, -1.0, 1.0]]) << u.m
@@ -2274,6 +2281,17 @@ class TestLinAlg(metaclass=CoverageMeta):
         w = np.linalg.eigvalsh(self.q)
         wx = np.linalg.eigvalsh(self.q.value) << self.q.unit
         assert_array_equal(w, wx)
+
+    if not NUMPY_LT_2_0:
+        # Numpy 2.0 added array-api compatible definitions of
+        # diagonal and trace to np.linalg. Since these have
+        # name conflicts with the main numpy namespace, they
+        # are tracked as linalg_diagonal and linalg_trace.
+        def test_linalg_diagonal(self):
+            self.check(np.linalg.diagonal)
+
+        def test_linalg_trace(self):
+            self.check(np.trace)
 
 
 class TestRecFunctions(metaclass=CoverageMeta):

